agent/engine/docker_task_engine.go

```diff
@@ -245,53 +245,73 @@ func (engine *DockerTaskEngine) synchronizeState() {
 	}
 
 	tasks := engine.state.AllTasks()
+	var tasksToStart []*api.Task
 	for _, task := range tasks {
 		conts, ok := engine.state.ContainerMapByArn(task.Arn)
 		if !ok {
-			engine.startTask(task)
+			// task hasn't start processing, no need to check container status
+			tasksToStart = append(tasksToStart, task)
 			continue
 		}
+
 		for _, cont := range conts {
-			if cont.DockerID == "" {
-				log.Debug("Found container potentially created while we were down", "name", cont.DockerName)
-				// Figure out the dockerid
-				describedCont, err := engine.client.InspectContainer(cont.DockerName, inspectContainerTimeout)
-				if err != nil {
-					log.Warn("Could not find matching container for expected", "name", cont.DockerName)
-				} else {
-					cont.DockerID = describedCont.ID
-					cont.Container.SetKnownStatus(dockerStateToState(describedCont.State))
-					// update mappings that need dockerid
-					engine.state.AddContainer(cont, task)
-					engine.imageManager.RecordContainerReference(cont.Container)
-				}
-			}
-			if cont.DockerID != "" {
-				currentState, metadata := engine.client.DescribeContainer(cont.DockerID)
-				if metadata.Error != nil {
-					currentState = api.ContainerStopped
-					if !cont.Container.KnownTerminal() {
-						cont.Container.ApplyingError = api.NewNamedError(&ContainerVanishedError{})
-						log.Warn("Could not describe previously known container; assuming dead", "err", metadata.Error, "id", cont.DockerID, "name", cont.DockerName)
-						engine.imageManager.RemoveContainerReferenceFromImageState(cont.Container)
-					}
-				} else {
-					engine.imageManager.RecordContainerReference(cont.Container)
-					if engine.cfg.ContainerMetadataEnabled && !cont.Container.IsMetadataFileUpdated() {
-						go engine.updateMetadataFile(task, cont)
-					}
-
-				}
-				if currentState > cont.Container.GetKnownStatus() {
-					cont.Container.SetKnownStatus(currentState)
-				}
-			}
+			engine.synchronizeContainerStatus(cont, task)
+		}
+
+		tasksToStart = append(tasksToStart, task)
+
+		// Put tasks that are stopped by acs but hasn't been stopped in wait group
+		if task.GetDesiredStatus().Terminal() && task.GetStopSequenceNumber() != 0 {
+			engine.taskStopGroup.Add(task.GetStopSequenceNumber(), 1)
 		}
+	}
+
+	for _, task := range tasksToStart {
 		engine.startTask(task)
 	}
+
 	engine.saver.Save()
 }
 
+// synchronizeContainerStatus checks and updates the container status with docker
+func (engine *DockerTaskEngine) synchronizeContainerStatus(container *api.DockerContainer, task *api.Task) {
+	if container.DockerID == "" {
+		log.Debug("Found container potentially created while we were down", "name", container.DockerName)
+		// Figure out the dockerid
+		describedContainer, err := engine.client.InspectContainer(container.DockerName, inspectContainerTimeout)
+		if err != nil {
+			log.Warn("Could not find matching container for expected", "name", container.DockerName)
```

Comment:
What happens to the task at this point of time?

Reply:
Nothing will happen at this point, task won't update the status of this container for this cycle of synchronizing. It looks like the behavior is a little different for containers with dockerID and containers without dockerID. I have created an issue for this, based on your comments below.